Fix VSO 487701 and 487702.

pgavlin opened the pull request:
Both of these issues stem from attempting to unassign a copied interval
from the copied-to register and then reassigning the interval to the
same. This corrupts some of the bookkeeping necessary to track whether
or not the interval in the register needs to be spilled, and the RA ends
up attempting to spill the interval being allocated even though it is
not assigned a register.

pgavlin commented:
@CarolEidt @sandreenko @sjsinju @dotnet/jit-contrib PTAL

sandreenko commented:
If it not hard, could you do PR with tests and then commit the fix? I want to see ci jobs results with tests before the fix.

pgavlin commented:
> If it not hard, could you do PR with tests and then commit the fix? I want to see ci jobs results with tests before the fix.

Why? Just to prove that they fail without this fix?

sandreenko commented on src/jit/lsra.cpp:
```diff
@@ -5723,7 +5723,7 @@ regNumber LinearScan::tryAllocateFreeReg(Interval* currentInterval, RefPosition*
-        if (isAssigned(availablePhysRegInterval ARM_ARG(currentInterval->registerType)))
+        if ((intervalToUnassign != nullptr) && isAssigned(availablePhysRegInterval ARM_ARG(currentInterval->registerType)))
         {
             intervalToUnassign = availablePhysRegInterval->assignedInterval;
```
As I see we did not use the value of `intervalToUnassign ` before. We always rewrite it with `availablePhysRegInterval->assignedInterval` or did not use at all.

Can we replace it with like `bool needUnassign = false` use it in the condition and declare 'intervalToUnassign' in this if.

pgavlin replied:
> As I see we did not use the value of `intervalToUnassign` before. We always rewrite it with `availablePhysRegInterval->assignedInterval` or did not use at all.

`intervalToUnassign` was used before the changes in https://github.com/dotnet/coreclr/commit/62ef107532203badddded0cc7f7a06765afffa7a which introduced this regression.

> Can we replace it with like `bool needUnassign = false` use it in the condition and declare `intervalToUnassign` in this if.

Sure, that's fine.

sandreenko commented:
> > If it not hard, could you do PR with tests and then commit the fix? I want to see ci jobs results with tests before the fix.
> 
> Why? Just to prove that they fail without this fix?

To see which jobs will fail. I am just curious will we see it for x86 non-windows or not.

pgavlin commented:
> To see which jobs will fail. I am just curious will we see it for x86 non-windows or not.

Alright, I've kicked off jobs for b426e68.

sandreenko commented:
Hm, why have all tests passed? 
New tests are pri1, but I think `Windows_NT x64 Release Priority 1 Build and Test` should fail, because it is not only assert, but null pointer exception.

Also how do we chose pri0 or pri1 for new tests?

pgavlin commented:
> Hm, why have all tests passed?

Yeah, I noticed this. Not sure. They fail when run locally...

FWIW, I added them as Pri0 in the first commit and changed them to Pri1 with the second.

> Also how do we chose pri0 or pri1 for new tests?

That is a good question that I can't yet answer satisfactorily. I've been leaning towards adding new tests as Pri1 with the thought that we can move them to Pri0 if they prove to be valuable indicators of new issues.

pgavlin commented:
I wonder if something went a bit funny w.r.t. what code was getting built. I may have pushed the commit with the fix before the only leg that I am certain should have failed started (i.e. `Windows_NT x64 Debug`). In any case, these tests did fail when run locally and no longer fail with the fix. PTAL.

sandreenko commented:
> I wonder if something went a bit funny w.r.t. what code was getting built. I may have pushed the commit with the fix before the only leg that I am certain should have failed started (i.e. Windows_NT x64 Debug). In any case, these tests did fail when run locally and no longer fail with the fix. PTAL.

Windows_NT x64 Debug doesn't run Pri1 tests, so it was not executed at all (you can search for test name in the [log ](https://ci.dot.net/job/dotnet_coreclr/job/master/job/debug_windows_nt_prtest/10599/consoleText)to check).

It was built and ran in Release mode, but [passed](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_release_windows_nt_pri1_prtest/10542/testReport/JIT_Regression/_JitBlue_DevDiv_487701_DevDiv_487701_DevDiv_487701_/).

pgavlin commented:
> Windows_NT x64 Debug doesn't run Pri1 tests, so it was not executed at all (you can search for test name in the log to check).

I'm aware. As I mentioned, the first commit added the tests as Pri0 tests in the hopes of avoiding this issue. The second commit moved them to Pri1 (hence my hypothesis regarding a race).
